### DISCUSSION TITLE
Upcoming changes in pypdfium2 & suggestions to improve PDF rendering integration

### DISCUSSION THREAD

**Discussion by mara004** [mindee/doctr#844] [Ideas]
Hello,

I'm the maintainer of pypdfium2 and found your interesting library using [sourcegraph](https://sourcegraph.com/search?q=context:global+pypdfium2&patternType=literal). It's amazing to see pypdfium2 in practical use!

The PDF backend code of `doctr` seems to be defined in [`io/pdf.py`](https://github.com/mindee/doctr/blob/main/doctr/io/pdf.py). I took a quick look and have some considerations:
* pypdfium2 does not support file paths as `pathlib.Path` anymore. I initially had support for pathlib but then switched back to `os.path` for various reasons. This means you should probably call the rendering function with `str(file)` if `file` is a path.
* At the time your PDF backend was implemented, the support model of pypdfium2 could only return `PIL.Image.Image`. I recently changed the API to allow getting the data as bytes: `render_pdf()` is replaced by `render_pdf_topil()` and `render_pdf_tobytes()`.
* nit: There is a code comment that still assumes use of pymupdf: `# Read pages with fitz and convert them to numpy ndarrays`

Kind regards

**Comment by fg-mindee** [comment]
Hi @mara004 :wave: 

Thanks a lot for the heads up!
- casting it to a string is a very viable option :+1: 
- glad the library keeps on improving! I suggest waiting for your next stable release on pypi to make this change :)
- good catch!

Happy to to discuss this further if needed!
